src/deluge/gui/views/midi_session_view.cpp

```diff
@@ -0,0 +1,949 @@
+				//init knobPos for current param
+				previousKnobPos[lastPadPress.xDisplay][lastPadPress.yDisplay] = kNoSelection;
+
+				//look to see if this cc was mapped elsewhere and we have a previousKnobPosition saved
+				for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+					for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+						if ((paramToCC[xDisplay][yDisplay] == ccNumber)
+						    && (previousKnobPos[xDisplay][yDisplay] != kNoSelection)) {
+							previousKnobPos[lastPadPress.xDisplay][lastPadPress.yDisplay] =
+							    previousKnobPos[xDisplay][yDisplay];
+						}
+					}
+				}
+
+				//assign cc to current param selected
+				paramToCC[lastPadPress.xDisplay][lastPadPress.yDisplay] = ccNumber;
+
+				renderParamDisplay(lastPadPress.paramKind, lastPadPress.paramID, ccNumber);
+
+				currentCC = kNoSelection;
+				updateMappingChangeStatus();
+			}
+		}
+		else {
+			currentCC = ccNumber;
+		}
+		uiNeedsRendering(this);
+	}
+	else if (lastPadPress.isActive) {
+		cantLearn(channel);
+	}
+}
+
+void MidiSessionView::cantLearn(int32_t channel) {
+	if (display->haveOLED()) {
+		char cantBuffer[40] = {0};
+		strncat(cantBuffer, l10n::get(l10n::String::STRING_FOR_CANT_LEARN), 39);
+		strncat(cantBuffer, l10n::get(l10n::String::STRING_FOR_MIDI_LEARN_CHANNEL), 39);
+
+		char buffer[5];
+		intToString(channel + 1, buffer);
+
+		strncat(cantBuffer, buffer, 4);
+
+		display->displayPopup(cantBuffer);
+	}
+	else {
+		display->displayPopup(l10n::get(l10n::String::STRING_FOR_CANT_LEARN));
+	}
+}
+
+void MidiSessionView::selectEncoderAction(int8_t offset) {
+	return;
+}
+
+ActionResult MidiSessionView::horizontalEncoderAction(int32_t offset) {
+	return ActionResult::DEALT_WITH;
+}
+
+ActionResult MidiSessionView::verticalEncoderAction(int32_t offset, bool inCardRoutine) {
+	return ActionResult::DEALT_WITH;
+}
+
+/// why do I need this? (code won't compile without it)
+uint32_t MidiSessionView::getMaxZoom() {
+	return currentSong->getLongestClip(true, false)->getMaxZoom();
+}
+
+/// why do I need this? (code won't compile without it)
+uint32_t MidiSessionView::getMaxLength() {
+	return currentSong->getLongestClip(true, false)->loopLength;
+}
+
+void MidiSessionView::modEncoderAction(int32_t whichModEncoder, int32_t offset) {
+	return;
+}
+
+/// used to reset stutter if it's already active
+void MidiSessionView::modEncoderButtonAction(uint8_t whichModEncoder, bool on) {
+	return;
+}
+
+void MidiSessionView::modButtonAction(uint8_t whichButton, bool on) {
+	UI::modButtonAction(whichButton, on);
+}
+
+ModelStackWithAutoParam* MidiSessionView::getModelStackWithParam(int32_t xDisplay, int32_t yDisplay, int32_t ccNumber,
+                                                                 bool displayError) {
+	ModelStackWithAutoParam* modelStackWithParam = nullptr;
+	if (currentSong) {
+		Param::Kind paramKind = Param::Kind::NONE;
+		int32_t paramID = kNoParamID;
+		char modelStackMemory[MODEL_STACK_MAX_SIZE];
+
```

Comment:
add description of what this code is doing - it's seeing if it can get a clip from the current context